internal/handlers/pg/pgdb/query_iterator.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pgdb
+
+import (
+	"context"
+	"runtime"
+	"sync"
+	"sync/atomic"
+
+	"github.com/jackc/pgx/v4"
+
+	"github.com/FerretDB/FerretDB/internal/handlers/pg/pjson"
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/iterator"
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
+)
+
+// queryIterator implements iterator.Interface to fetch documents from the database.
+type queryIterator struct {
+	ctx         context.Context
+	rows        pgx.Rows
+	currentIter atomic.Uint32
+	closed      bool        // indicates whether Close() was called
+	mxClosed    *sync.Mutex // protects closed
+}
+
+// newIterator returns a new queryIterator for the given pgx.Rows.
+// It sets finalizer to close the rows.
+func newIterator(ctx context.Context, rows pgx.Rows) iterator.Interface[uint32, *types.Document] {
+	// queryIterator is defined as pointer to address it in the finalizer.
+	qi := &queryIterator{
+		ctx:      ctx,
+		rows:     rows,
+		closed:   false,
+		mxClosed: new(sync.Mutex),
+	}
+
+	runtime.SetFinalizer(qi, func(qi *queryIterator) {
+		qi.mxClosed.Lock()
+		defer qi.mxClosed.Unlock()
+		if !qi.closed {
+			// TODO: ask about fatalpanic
+			panic("queryIterator.Close() has not been called")
```

Comment:
Important: this panic is `fatalpanic` (unrecoverable) as it's inside the finalizer, and there is no way to catch the function that would execute it. So, if it happens, FerretDB stops, `recover` will be ignored.

Reply:
> Are we ok with keeping this panic?

Yes